### PR TITLE
Fixes #24474

### LINKED ISSUE #24474: There is a memory leak defect at line 141 in the file /openssl/test/cmp_hdr_test.c.
This defect is located in a different function than the one mentioned in the issue https://github.com/openssl/openssl/issues/24469. They are two distinct defects, even though they are in the same file and the variable names are the same.
In the file /openssl/test/cmp_hdr_test.c, at line 134, a pointer variable named x509name is defined and allocated a dynamic memory region through the function X509_NAME_new. When the if statement at line 136 returns false, it indicates that the dynamic memory region pointed to by x509name was successfully allocated. If the if statement at line 140 returns true, the program will return at line 141, thereby skipping the release of the dynamic memory region pointed to by x509name at line 150. This results in a memory leak defect, as shown in the following diagram(note that the defect path provided in the figure does not match my description; my description should be considered accurate):
https://github.com/LuMingYinDetect/openssl_defects/blob/main/openssl_28.png

### PR DESCRIPTION
**Pull request by cm-jones**
Fixes #24474

### CODE REVIEW

File: test/cmp_hdr_test.c [hunk 1]
```diff
@@ -137,8 +137,10 @@ static int execute_HDR_set1_recipient_test(CMP_HDR_TEST_FIXTURE *fixture)
         return 0;
 
     X509_NAME_ADD(x509name, "CN", "A common recipient name");
-    if (!TEST_int_eq(ossl_cmp_hdr_set1_recipient(fixture->hdr, x509name), 1))
+    if (!TEST_int_eq(ossl_cmp_hdr_set1_recipient(fixture->hdr, x509name), 1)) {
+        X509_NAME_free(x509name);
```
Comment (nhorman) [inline test/cmp_hdr_test.c:141]: Doesnt this have to be ddone for evvery return zero in this function?  i.e. perhaps you should make the return zeros a goto out, where the X509_NAME_free is already called, and then just return a stack variable properly set to 1 or 0?

**Review by nhorman — CHANGES_REQUESTED** [review]

File: test/cmp_hdr_test.c (hunk 1 above)
Comment (cm-jones) [reply test/cmp_hdr_test.c:141]: Like this?

```
static int execute_HDR_set1_recipient_test(CMP_HDR_TEST_FIXTURE *fixture)
{
    X509_NAME *x509name = X509_NAME_new();
    int res = 0;

    if (!TEST_ptr(x509name))
        return 0;

    X509_NAME_ADD(x509name, "CN", "A common recipient name");
    if (!TEST_int_eq(ossl_cmp_hdr_set1_recipient(fixture->hdr, x509name), 1))
        goto err;

    if (!TEST_int_eq(fixture->hdr->recipient->type, GEN_DIRNAME))
        goto err;

    if (!TEST_int_eq(X509_NAME_cmp(fixture->hdr->recipient->d.directoryName,
                                   x509name), 0))
        goto err;

    res = 1;

 err:
    X509_NAME_free(x509name);
    return res;
}
```

**Review by t8m — APPROVED** [review]

**Comment by cm-jones** [comment]
@t8m Submitted a signed ICLA to legal@opensslfoundation.org.

**Comment by cm-jones** [comment]
Done.

**Comment by t8m** [comment]
Ah, there is a rebase conflict - so please update the master branch from openssl/openssl repo and do `git rebase master` to rebase and resolve the conflict.

**Review by shahsb — APPROVED** [review]
LGTM. Thanks!

**Comment by cm-jones** [comment]
Was the rebase successful? Apologies. I'm new to open source.

**Comment by nhorman** [comment]
I'm afraid the rebase doesn't look right to me, the last commiit you have on this branch  is "Merge branch 'master' into fix-24474" and its quite large.

If you're looking for guidance on how to do a rebase, start with your initial two commits, and from there:
1) git remote update
2) git pull <upstream remove> master:master
3) git checkout fix-24474
4) git rebase master fix-24474

That last step will require manual intervention, as we know there is a conflict.  git will stop on one of your two commits, telling you there is a conflict. you will need to manually edit the conflict to fix it at which point you will run:
1) git add <conflicted file>
2) git rebase --continue

The rebase should complete at that point, leaving your with your two commits, adjusted to apply cleanly to the head of the master branch.

**Comment by slontis** [comment]
If you click on the "files changed" above you will see what the applied changes are. It is obviously not right.
A more descriptive commit message would also be welcome. Such as
````
Fix memory leak in cmp_hdr_test
Fixes #24474
````

**Comment by tom-cosgrove-arm** [comment]
This PR seems to have collapsed down to duplicating the line `int res = 0` - in which case should it be closed?

**Pull request closed without merging**